power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset.hpp

```diff
@@ -45,8 +47,122 @@ struct DatasetInfo {
+
+    class iterator;
+
+    class Proxy {
+      public:
+        using value_type = T;
+
+        Proxy() = default;
+        Proxy(Idx idx, std::span<Data* const> data,
+              std::span<std::reference_wrapper<MetaAttribute const> const> meta_attributes)
+            : idx_{idx}, data_{data}, meta_attributes_{meta_attributes} {
+            assert(data.size() == meta_attributes.size());
+        }
+
+        Proxy& operator=(value_type const& value)
+            requires is_data_mutable_v<dataset_type>
+        {
+            for (Idx attribute_idx = 0; attribute_idx < static_cast<Idx>(meta_attributes_.size()); ++attribute_idx) {
+                auto const& meta_attribute = get_meta_attribute(attribute_idx);
+                std::byte* buffer_ptr = reinterpret_cast<std::byte*>(data_[attribute_idx]) + meta_attribute.size * idx_;
+                meta_attribute.get_value(&value, buffer_ptr, 0);
+            }
+            return *this;
+        }
+        operator value_type() const { return get(); }
+        value_type get() const {
+            std::remove_const_t<value_type> result{};
+            for (Idx attribute_idx = 0; attribute_idx < static_cast<Idx>(meta_attributes_.size()); ++attribute_idx) {
+                auto const& meta_attribute = get_meta_attribute(attribute_idx);
+                std::byte const* buffer_ptr =
+                    reinterpret_cast<std::byte const*>(data_[attribute_idx]) + meta_attribute.size * idx_;
+                meta_attribute.set_value(&result, buffer_ptr, 0);
+            }
+            return result;
+        }
+
+      private:
+        friend class ColumnarAttributeRange;
+        friend class iterator;
+
+        MetaAttribute const& get_meta_attribute(Idx attribute_idx) const {
+            return meta_attributes_[attribute_idx].get();
+        }
+
+        Idx idx_{};
+        std::span<Data* const> data_{};
+        std::span<std::reference_wrapper<MetaAttribute const> const> meta_attributes_;
+    };
+
+    class iterator : public boost::iterator_facade<iterator, T, boost::random_access_traversal_tag, Proxy, Idx> {
+      public:
+        using value_type = Proxy;
+
+        iterator() = default;
+        iterator(Idx idx, std::span<Data* const> data,
+                 std::span<std::reference_wrapper<MetaAttribute const> const> meta_attributes)
+            : current_{idx, data, meta_attributes} {}
+
+      private:
+        friend class boost::iterator_core_access;
+
+        constexpr auto dereference() const -> value_type { return *&current_; }
+        constexpr auto dereference() -> value_type { return *&current_; }
+        constexpr auto equal(iterator const& other) const {
+            return current_.idx_ == other.current_.idx_ &&
+                   current_.meta_attributes_.size() == other.current_.meta_attributes_.size() &&
+                   current_.meta_attributes_.begin() == other.current_.meta_attributes_.begin();
```

Comment:
These two comparison is confusing. Especially the comparing the `begin`. If the attributes are empty, this is actually UB.

We don't need to do this comparison. Just compare the index. (internal) Users are not supposed to compare two iterators from different range.

Reply:
done